### ISSUE TITLE
Proposal : Redux action structure

### ISSUE THREAD

**Issue by BenMaruchu** [CodeTanzania/emis-dashboard#56]
This is a proposal on how we can structure our redux actions

### Action types names

I propose that actions types to have this kind of naming pattern i.e `Resource(plural)_verb(http)_stage(name)`

e.g

- USERS_GET_START
- USERS_GET_SUCCESS
- USERS_GET_ERROR
- USERS_POST_START
- USERS_DELETE_START
- USERS_PUT_START

```js
{
    type:'USERS_GET_START',
    meta:{
        page:1,
        query:''
    }
}

...

{
    type:'USERS_GET_SUCCESS',
    payload:{
        data:''
    },
    meta:{
        page:1,
        query:''
    }
}

...

{
    type:'USERS_GET_ERROR',
    payload:{
        data:'' // error object
    },
    error:true
}

```

Then each action which is supposed to have data should have payload property that will carry data, and meta object which will have meta data about the data carried in the payload or meta data to be used on querying the API.

### References

Some of the references used on this proposal

[flux-standard-action](https://github.com/redux-utilities/flux-standard-action)

**Comment by BenMaruchu** [comment]
Guy please review the proposed structure and give out your views on this

**Comment by kasongoyo** [comment]
I like the concept and it will aid a lot on documentation about the actions. I think from the reference material, it's clear that one of the benefits of the proposed style is to get rid of `ACTION_SUCCESS` & `ACTION_ERROR` so if we opt for FSA then it's better we embrace this approach.

**Comment by BenMaruchu** [comment]
@kasongoyo But using pure FSA will be a little bit difficult cause we will have to write logics in reducers, which I don't think it is a wise thing to do.

**Comment by Mloweedgar** [comment]
I like the Proposed structure, I will start using it.

**Comment by kasongoyo** [comment]
@benson, yeah, when I said we should use pure FSA, I was not intendin that statement to be taken literally. I was just addresin issue of action success and action error. I like the idea of FSA that we should not have additional actions for handlin async of another action instead we should use the proposed action structure to tackle async result so instead of havin action error as you posted above then the combination of error key and payload in action can solve that.

**Comment by BenMaruchu** [comment]
I have a suggestion on the action type name. We should start with verbs instead of noun
